src/xas_tdp_atom.F

```diff
@@ -0,0 +1,3362 @@
+!>        target atom wrt to the position of source atom
+!> \param rho_set where the densities are stored
+!> \param ri_dcoeff the arrays containing the RI density coefficient of source_iat, for each spin
+!> \param source_iat the index of the source atom
+!> \param source_ikind the kind of the source atom
+!> \param target_iat the index of the target atom
+!> \param target_ikind the kind of the target atom
+!> \param do_gga whether the gradient of the density is needed
+!> \param bounds the local bounds of the atomic grid
+!> \param xas_atom_env ...
+!> \param qs_env ...
+! **************************************************************************************************
+   SUBROUTINE put_density_on_other_grid(rho_set, ri_dcoeff, source_iat, source_ikind, target_iat, &
+                                        target_ikind, do_gga, bounds, xas_atom_env, qs_env)
+
+      TYPE(xc_rho_set_type), POINTER                     :: rho_set
+      TYPE(cp_1d_r_p_type), DIMENSION(:)                 :: ri_dcoeff
+      INTEGER, INTENT(IN)                                :: source_iat, source_ikind, target_iat, &
+                                                            target_ikind
+      LOGICAL, INTENT(IN)                                :: do_gga
+      INTEGER, DIMENSION(2, 3), INTENT(IN)               :: bounds
+      TYPE(xas_atom_env_type), POINTER                   :: xas_atom_env
+      TYPE(qs_environment_type), POINTER                 :: qs_env
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'put_density_on_other_grid', &
+         routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: dir, er, handle, ia, ico, ipgf, ir, &
+                                                            iset, isgf, lx, ly, lz, n, na, nr, &
+                                                            nset, nspins, sgfi, sr, start
+      INTEGER, DIMENSION(:), POINTER                     :: lmax, lmin, npgf, nsgf_set
+      INTEGER, DIMENSION(:, :), POINTER                  :: first_sgf
+      REAL(dp), ALLOCATABLE, DIMENSION(:, :)             :: sgf
+      REAL(dp), ALLOCATABLE, DIMENSION(:, :, :)          :: co, dsgf, pos
+      REAL(dp), ALLOCATABLE, DIMENSION(:, :, :, :)       :: dco
+      REAL(dp), DIMENSION(3)                             :: rs, rst, rt
+      REAL(dp), DIMENSION(:, :), POINTER                 :: ri_sphi, zet
+      REAL(dp), DIMENSION(:, :, :), POINTER              :: rhoa, rhob
+      TYPE(cell_type), POINTER                           :: cell
+      TYPE(grid_atom_type), POINTER                      :: grid_atom
+      TYPE(gto_basis_set_type), POINTER                  :: ri_basis
+      TYPE(harmonics_atom_type), POINTER                 :: harmonics
+      TYPE(particle_type), DIMENSION(:), POINTER         :: particle_set
+      TYPE(qs_kind_type), DIMENSION(:), POINTER          :: qs_kind_set
+
+      NULLIFY (qs_kind_set, r
```

Comment:
should be a `parameter`, possibly factor out to `src/common/physcon.F` together with `velit` from `src/dkh_main.F`

Reply:
Done. Also refactored `cspeed` from `atom_operators.F`